### PR TITLE
[No QA][TS migration] Migrate 'CheckForPreviousReportActionID.js' lib to TypeScript

### PR DESCRIPTION
**Pull request by fvlvte**
### Details
Migrated CheckForPreviousReportActionID.js and it's dependencies to TypeScript.
Aligned code to guidelines and ESLint rules.

### Fixed Issues
$ https://github.com/Expensify/App/issues/24856
PROPOSAL: N/A

### Tests
- [x] Verify that no errors appear in the JS console

### Offline tests
N/A

### QA Steps
- [x] Verify that no errors appear in the JS console

### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/20582108/b8d90d83-32f2-4513-84d1-9ab5cd09564f

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/20582108/3a5bdd16-b77b-4d3c-8e65-084aa58b937e

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/20582108/a58c97d4-753c-4fda-b771-5f3c783fc150

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/20582108/4225da9d-7aa2-4339-81b4-a6963372707d

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/20582108/26c64373-ec1a-40cb-8633-703ecd4d18b7

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/20582108/d0725a81-7a76-4a2a-a352-cae3d9bea5a2

</details>

### CODE REVIEW

**Comment by fabioh8010** [comment]
Let's hold this one until https://github.com/Expensify/react-native-onyx/pull/353 is merged, we need to be able to pass `null` without getting type errors.

**Comment by fabioh8010** [comment]
Also @fvlvte Please rebase and sign your old commits that are unverified.

**Comment by blazejkustra** [comment]
@fvlvte No longer blocked

**Review by kubabutkiewicz — APPROVED** [review]
lgtm 👍

File: src/libs/migrations/CheckForPreviousReportActionID.ts [hunk 1]
```diff
@@ -44,24 +40,25 @@ export default function () {
...
-        const onyxData = {};
-        _.each(allReportActions, (reportAction, onyxKey) => {
+        const onyxData: OnyxCollection<OnyxTypes.ReportActions> = {};
+
+        Object.keys(allReportActions ?? {}).forEach((onyxKey) => {
             onyxData[onyxKey] = {};
         });
 
-        return Onyx.multiSet(onyxData);
+        return Onyx.multiSet(onyxData as Record<`${typeof ONYXKEYS.COLLECTION.REPORT_ACTIONS}`, Record<string, never>>);
```
Comment (VickyStash) [inline src/libs/migrations/CheckForPreviousReportActionID.ts:62]: @fvlvte Maybe we can update the `Onyx.multiSet` type to get rid of the assertion?

File: src/ONYXKEYS.ts [hunk 2]
```diff
@@ -386,7 +386,7 @@ type OnyxValues = {
     [ONYXKEYS.COLLECTION.WORKSPACE_INVITE_MEMBERS_DRAFT]: Record<string, number>;
     [ONYXKEYS.COLLECTION.REPORT]: OnyxTypes.Report;
     [ONYXKEYS.COLLECTION.REPORT_METADATA]: OnyxTypes.ReportMetadata;
-    [ONYXKEYS.COLLECTION.REPORT_ACTIONS]: OnyxTypes.ReportAction;
+    [ONYXKEYS.COLLECTION.REPORT_ACTIONS]: OnyxTypes.ReportActions;
```
Comment (blazejkustra) [inline src/ONYXKEYS.ts:389]: This change is already on the newest main, so please merge main here

File: src/libs/migrations/CheckForPreviousReportActionID.ts [hunk 3]
```diff
@@ -1,17 +1,14 @@
...
- * @returns {Promise<Object>}
- */
-function getReportActionsFromOnyx() {
+function getReportActionsFromOnyx(): Promise<OnyxCollection<OnyxTypes.ReportActions>> {
     return new Promise((resolve) => {
         const connectionID = Onyx.connect({
             key: ONYXKEYS.COLLECTION.REPORT_ACTIONS,
             waitForCollectionCallback: true,
-            callback: (allReportActions) => {
+            callback: (allReportActions: OnyxCollection<OnyxTypes.ReportActions>) => {
```
Comment (blazejkustra) [inline src/libs/migrations/CheckForPreviousReportActionID.ts:11]: No need for the type:
```suggestion
            callback: (allReportActions) => {
```

File: src/types/onyx/ReportAction.ts [hunk 4]
```diff
@@ -85,8 +85,12 @@ type ReportActionBase = {
     childVisibleActionCount?: number;
 
     pendingAction?: OnyxCommon.PendingAction;
+    errors?: OnyxCommon.Errors;
```
Comment (blazejkustra) [inline src/types/onyx/ReportAction.ts:88]: Add a comment explaining the type

**Review by blazejkustra — CHANGES_REQUESTED** [review]

File: src/libs/migrations/CheckForPreviousReportActionID.ts [hunk 5]
```diff
@@ -22,20 +19,19 @@ function getReportActionsFromOnyx() {
...
 
-        let firstValidValue;
-        _.some(_.values(allReportActions), (reportActions) =>
-            _.some(_.values(reportActions), (reportActionData) => {
-                if (_.has(reportActionData, 'reportActionID')) {
+        let firstValidValue: OnyxTypes.ReportAction | undefined;
+
+        Object.values(allReportActions ?? {}).some((reportActions) =>
+            Object.values(reportActions ?? {}).some((reportActionData) => {
+                if (reportActionData?.reportActionID) {
```
[the remark this answers is not in the document]
Comment (blazejkustra) [reply src/libs/migrations/CheckForPreviousReportActionID.ts]: Logically `reportActionData.reportActionID?.length` is the same as `reportActionData?.reportActionID`. This condition is true when string is not empty.

Before this PR it worked differently, condition was true even for empty strings, that's why I suggested `'reportActionID' in reportActionData`

**Review by blazejkustra — APPROVED** [review]

**Review by blazejkustra — APPROVED** [review]

**Review by thesahindia — APPROVED** [review]

File: src/libs/migrations/CheckForPreviousReportActionID.ts [hunk 6]
```diff
@@ -44,24 +40,25 @@ export default function () {
...
         );
 
-        if (_.isUndefined(firstValidValue)) {
+        if (!firstValidValue) {
             Log.info(`[Migrate Onyx] Skipped migration CheckForPreviousReportActionID because there were no valid reportActions`);
             return;
         }
 
-        if (_.has(firstValidValue, 'previousReportActionID')) {
+        if (firstValidValue.previousReportActionID) {
```
Comment (dangrous) [inline src/libs/migrations/CheckForPreviousReportActionID.ts:48]: will previousReportActionID ever be 0 or otherwise falsey?

**Review by dangrous — COMMENTED** [review]
One question and then can you explain why the switch from integers to strings in the test? Is some of the functionality changing? It looks fine otherwise. Thanks!

**Comment by dangrous** [comment]
Hi! Following up on the above questions when you have a moment - thnaks! Then we can get this merged

**Comment by dangrous** [comment]
okay great! And why did the tests change from integers to strings? Since it's passing I'm guessing it's fine but I wanted to make sure I understood.

**Review by dangrous — APPROVED** [review]
LGTM!

**Pull request merged**